File: Java/src/main/java/QuestionBank/Solution601_700/Solution606.java
package QuestionBank.Solution601_700;

import tool.TreeGenerate;
import tool.TreeNode;

/**
 * @author windows
 */
public class Solution606 {
    StringBuilder sb = new StringBuilder();
    public String tree2str(TreeNode root) {
        if (root == null){
            return sb.toString();
        }
        sb.append(root.val);
        if (root.left != null || root.right!= null) {
            sb.append('(');
            tree2str(root.left);
            sb.append(')');
            sb.append('(');
            tree2str(root.right);
            sb.append(')');
        }
        String replace = sb.toString().replace("())", ")");
        if (replace.endsWith("()")){
            replace = replace.substring(0, replace.length() - 2);
        }
        return replace;
    }



    public static void main(String[] args) {
        Solution606 solution606 = new Solution606();
        TreeNode tree = TreeGenerate.createTree(new Integer[]{1,2,null,3,4});
        System.out.println(solution606.tree2str(tree));
    }
}

File: C++/src/Solution601_700/Solution606.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution606.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;


//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


void dfs(TreeNode* root, string &s){
  if (root == nullptr){
    return;
  }
  s += to_string(root->val);
  if (root->left != nullptr){
    s += '(';
    dfs(root->left, s);
    s += ')';
  }
  if (root->left == nullptr && root->right != nullptr){
    s += "()";
  }
  if (root->right != nullptr){
    s += '(';
    dfs(root->right, s);
    s += ')';
  }
}

string tree2str(TreeNode* root) {
    string s = "";
    dfs(root, s);
    return s;
}
